src/daft-table/src/probeable/probe_set.rs

```diff
@@ -0,0 +1,170 @@
+use std::{
+    collections::{hash_map::RawEntryMut, HashMap},
+    sync::Arc,
+};
+
+use common_error::DaftResult;
+use daft_core::{
+    array::ops::as_arrow::AsArrow,
+    prelude::SchemaRef,
+    utils::{
+        dyn_compare::{build_dyn_multi_array_compare, MultiDynArrayComparator},
+        identity_hash_set::IdentityBuildHasher,
+    },
+};
+
+use super::{ArrowTableEntry, IndicesMapper, Probeable, ProbeableBuilder};
+use crate::{ops::hash::IndexHash, Table};
+pub(crate) struct ProbeSet {
+    schema: SchemaRef,
+    hash_table: HashMap<IndexHash, (), IdentityBuildHasher>,
+    tables: Vec<ArrowTableEntry>,
+    compare_fn: MultiDynArrayComparator,
+    num_groups: usize,
+    num_rows: usize,
+}
+
+impl ProbeSet {
+    // Use the leftmost 28 bits for the table index and the rightmost 36 bits for the row number
+    const TABLE_IDX_SHIFT: usize = 36;
+    const LOWER_MASK: u64 = (1 << Self::TABLE_IDX_SHIFT) - 1;
+
+    const DEFAULT_SIZE: usize = 20;
+
+    pub(crate) fn new(schema: SchemaRef) -> DaftResult<Self> {
+        let hash_table = HashMap::<IndexHash, (), IdentityBuildHasher>::with_capacity_and_hasher(
+            Self::DEFAULT_SIZE,
+            Default::default(),
+        );
+        let compare_fn = build_dyn_multi_array_compare(&schema, false, false)?;
+        Ok(Self {
+            schema,
+            hash_table,
+            tables: vec![],
+            compare_fn,
+            num_groups: 0,
+            num_rows: 0,
+        })
+    }
+
+    fn probe<'a>(&'a self, right: &'a Table) -> DaftResult<impl Iterator<Item = bool> + 'a> {
+        assert_eq!(self.schema.len(), right.schema.len());
+        assert!(self
+            .schema
+            .fields
+            .values()
+            .zip(right.schema.fields.values())
+            .all(|(l, r)| l.dtype == r.dtype));
+
+        let r_hashes = right.hash_rows()?;
+
+        let right_arrays = right
+            .columns
+            .iter()
+            .map(|s| Ok(s.as_physical()?.to_arrow()))
+            .collect::<DaftResult<Vec<_>>>()?;
+
+        let iter = r_hashes.as_arrow().clone().into_iter();
+
+        Ok(iter.enumerate().map(move |(r_idx, h)| {
+            if let Some(h) = h {
+                self.hash_table.raw_entry().from_hash(h, |other| {
+                    h == other.hash && {
+                        let l_idx = other.idx;
+                        let l_table_idx = (l_idx >> Self::TABLE_IDX_SHIFT) as usize;
+                        let l_row_idx = (l_idx & Self::LOWER_MASK) as usize;
+
+                        let l_table = self.tables.get(l_table_idx).unwrap();
+
+                        let left_refs = l_table.0.as_slice();
+
+                        (self.compare_fn)(left_refs, &right_arrays, l_row_idx, r_idx).is_eq()
```

Comment:
There's some kind of complicated logic here that's also repeated in `add_table`. It would be nice to have a better abstraction over it